### PR TITLE
Minkowski_sim_2: fix missed hole

### LINKED ISSUE #5830: The Minkowski Sum results using minkowski_sum_by_reduced_convolution_2() lose holes in some rare cases
## Issue Details

We demonstrate (possibly) a bug of Minkowski Sum Calculation using `minkowski_sum_by_reduced_convolution_2()` implemented in CGAL 5.2.2. We construct a `Polygon_with_holes_2` and perform Minkowski sum calculation using different padding values. The results show that if the padding values are set to 0.0001 or 1.0, the results are correct. However, when the padding value is set to 0.001, the result is wrong. By visualizing the padded result, it is easy to find that the center space of the original `Polygon_with_holes_2` wrongly becomes part of the padded result. In all cases, the debug mode does not report any issues, but the result is wrong when the padding value is set to 0.001.

## Source Code

```
#include <iostream>
#include <string>
#include <vector>

#include <CGAL/Gmpq.h>
#include <CGAL/Lazy_exact_nt.h>
#include <CGAL/Polygon_2.h>
#include <CGAL/Polygon_with_holes_2.h>
#include <CGAL/Simple_cartesian.h>
#include <CGAL/draw_polygon_with_holes_2.h>
#include <CGAL/minkowski_sum_2.h>

using KernelEE = CGAL::Simple_cartesian<CGAL::Lazy_exact_nt<CGAL::Gmpq>>;
using NumberEE = KernelEE::FT;
using Point2EE = KernelEE::Point_2;
using PolygonEE = CGAL::Polygon_2<KernelEE>;
using PolygonWHEE = CGAL::Polygon_with_holes_2<KernelEE>;

PolygonWHEE GetPolygonWHData() {
  auto StrToEE = [](const std::string& str, int base = 36) -> NumberEE {
    mpq_t mpq;
    mpq_init(mpq);
    mpq_set_str(mpq, str.c_str(), base);
    return NumberEE(std::move(mpq));
  };

  // step 1: construct the string data representing points
  std::vector<std::pair<std::string, std::string>> data;
  data.emplace_back("7hvid/4zsow", "1vhkh/18y68");
  data.emplace_back("87bxd/pa8", "7hvid/4zsow");
  data.emplace_back("87bxn/pa8", "5tna9/pa8");
  data.emplace_back("5hf9n/2t4w", "1getn/6bk");
  data.emplace_back("5hp01/2t4w", "4zn9r/pa8");
  data.emplace_back("7yvhx/2t4w", "8th2b/1ekg");
  data.emplace_back("12n6x/cn4", "44qcb/pa8");
  data.emplace_back("tdz9/35s", "89jxd/1ekg");
  data.emplace_back("1mryh/6bk", "qv71/cn4");
  data.emplace_back("21o0j/pa8", "5yxl3/2t4w");
  data.emplace_back("g46l/6bk", "57snn/2t4w");
  data.emplace_back("3zur3/1ekg", "6ybn/35s");
  data.emplace_back("1zwu9/pa8", "8s5hj/1ekg");
  data.emplace_back("8087/35s", "4lxcn/pa8");
  data.emplace_back("2ehdl/1ekg", "19mhh/6bk");
  data.emplace_back("3qynl/2hwcg", "5cou7/pa8");

  // step 2: convert from string data to PolygonWHEE
  PolygonEE polygon;
  for (const auto& pair : data)
    polygon.push_back(Point2EE(StrToEE(pair.first), StrToEE(pair.second)));
  return PolygonWHEE(polygon);
}

PolygonWHEE ExteriorPaddingOfPolygonWHEE(const PolygonWHEE& polygonwh, double padding) {
  std::vector<Point2EE> points{
      {-padding, padding}, {-padding, -padding}, {padding, -padding}, {padding, padding}};
  PolygonEE polygon(points.begin(), points.end());  // it is a square
  return CGAL::minkowski_sum_2(polygonwh, polygon);
}

void TestMinkowskiSumMissingHolesBug() {
  const PolygonWHEE original_polygonwh = GetPolygonWHData();

  // stpe 1: output the information of original polygonwh.
  // We check whether the outer boundary of the original polygonwh is simple,
  // and also output all the vertices of the outer boundary.
  {
    bool is_simple = original_polygonwh.outer_boundary().is_simple();
    std::cout << "Original polygonwh: " << std::endl;
    std::cout << "  outer boundary is simple? " << is_simple << std::endl;
    size_t index = 0;
    for (const Point2EE& point : original_polygonwh.outer_boundary())
      std::cout << "    " << index++ << ": " << point << std::endl;
    CGAL::draw(original_polygonwh);
  }

  // step 2: output the information of padded_polygonwh with 0.0001.
  // Since this padding value is small, the center region remains the
  // unbounded part of the padded polygonwh. Thus the result is correct.
  {
    PolygonWHEE padded_pwh = ExteriorPaddingOfPolygonWHEE(original_polygonwh, 0.0001);
    std::cout << "Padded polygonwh with 0.0001: " << std::endl;
    std::cout << "  the result is CORRECT." << std::endl;
    CGAL::draw(padded_pwh);
  }

  // step 3: output the information of padded_polygonwh with 0.001.
  // With this padding value, the center region becomes part of the padded
  // polygonwh. Thus the result is wrong.
  {
    PolygonWHEE padded_pwh = ExteriorPaddingOfPolygonWHEE(original_polygonwh, 0.001);
    std::cout << "Padded polygonwh with 0.001: " << std::endl;
    std::cout << "  the result is WRONG." << std::endl;
    CGAL::draw(padded_pwh);
  }

  // step 4: output the information of padded_polygonwh with 1.0.
  // Since this padding value is large, the center region becomes a hole of
  // the padded polygonwh. Thus the result is still correct.
  {
    PolygonWHEE padded_pwh = ExteriorPaddingOfPolygonWHEE(original_polygonwh, 1.0);
    std::cout << "Padded polygonwh with 1.0: " << std::endl;
    std::cout << " the result is CORRECT." << std::endl;
    CGAL::draw(padded_pwh);
  }

  std::cout << std::endl;
}

int main() {
  TestMinkowskiSumMissingHolesBug();
  return EXIT_SUCCESS;
}
```

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Windows 10 Professional, 64 bits / Linux (Ubuntu 20.04 LTS), 64 Bits
* Compiler: Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29334 for x64 on Windows; g++ 9.3.0-17 / clang 10.0.0-4 on Linux
* Release or debug mode: Release
* Specific flags used (if any):
* CGAL version: 5.2.2
* Boost version: 1.73.0
* Other libraries versions if used (Eigen, TBB, etc.):

## Screenshots

Original Polygon_with_holes_2: 
original

Padded Polygon_with_holes_2 with padding_offset = 0.0001 (correct):
padding_0 0001

Padded Polygon_with_holes_2 with padding_offset = 0.001 (wrong):
padding_0 001

Padded Polygon_with_holes_2 with padding_offset = 1.0 (correct):
padding_1 0

### PR DESCRIPTION
**Pull request by efifogel**
## Summary of Changes

Fixed a bug in Minkowski_sum_2.
In particular, fixed the function get_point_in_face()

## Release Management

* Affected package(s): Minkowski_sim_2
* Issue(s) solved (if any): fix #5830
* Feature/Small Feature (if any): NA
* Link to compiled documentation (obligatory for small feature): NA
* License and copyright ownership: Tel Aviv Univ.

### CODE REVIEW

File: Installation/CHANGES.md [hunk 1]
```diff
@@ -67,6 +67,9 @@ CGAL tetrahedral Delaunay refinement algorithm.
 -   The stop predicates `Count_stop_predicate` and `Count_ratio_stop_predicate` are renamed to `Edge_count_stop_predicate` and `Edge_count_ratio_stop_predicate`. Older versions have been deprecated.
 -   Introduce `Face_count_stop_predicate` and `Face_count_ratio_stop_predicate` that can be used to stop the simplification algorithm based on a desired number of faces in the output, or a ratio between input and output face numbers.
 
+### [2D Minkowski Sums](https://doc.cgal.org/5.6/Manual/packages.html#PkgMinkowskiSum2)
+-   Fixed a bug that mae holes in the Minkowski sum disappear
+
```
Comment (albert-github) [inline Installation/CHANGES.md:72]: mae?

File: Minkowski_sum_2/include/CGAL/Minkowski_sum_2/Minkowski_sum_by_reduced_convolution_2.h [hunk 2]
```diff
@@ -74,8 +73,8 @@ class Minkowski_sum_by_reduced_convolution_2
   typename Kernel::Counterclockwise_in_between_2 f_ccw_in_between;
 
 public:
-  Minkowski_sum_by_reduced_convolution_2()
-  {
+  //!
+  Minkowski_sum_by_reduced_convolution_2() {
```
Comment (albert-github) [inline Minkowski_sum_2/include/CGAL/Minkowski_sum_2/Minkowski_sum_by_reduced_convolution_2.h:77]: Why this extra comment line?

File: Minkowski_sum_2/include/CGAL/Minkowski_sum_2/Minkowski_sum_by_reduced_convolution_2.h [hunk 3]
```diff
@@ -86,10 +85,10 @@ class Minkowski_sum_by_reduced_convolution_2
     f_ccw_in_between = ker.counterclockwise_in_between_2_object();
   }
 
+  //!
```
Comment (albert-github) [inline Minkowski_sum_2/include/CGAL/Minkowski_sum_2/Minkowski_sum_by_reduced_convolution_2.h:88]: Why this extra comment line?

File: Minkowski_sum_2/include/CGAL/Minkowski_sum_2/Minkowski_sum_by_reduced_convolution_2.h [hunk 4]
```diff
@@ -109,23 +109,22 @@ class Minkowski_sum_by_reduced_convolution_2
     common_operator(pgn1, pgn2, outer_boundary, holes);
   }
 
+  //!
```
Comment (albert-github) [inline Minkowski_sum_2/include/CGAL/Minkowski_sum_2/Minkowski_sum_by_reduced_convolution_2.h:112]: Why this extra comment line?

File: Minkowski_sum_2/include/CGAL/Minkowski_sum_2/Minkowski_sum_by_reduced_convolution_2.h [hunk 5]
```diff
@@ -109,23 +109,22 @@ class Minkowski_sum_by_reduced_convolution_2
...
+  void operator()(const Polygon_2& pgn1, const Polygon_with_holes_2& pgn2,
+                  Polygon_2& outer_boundary, OutputIterator holes) const {
     CGAL_precondition(pgn1.is_simple());
     CGAL_precondition(pgn1.orientation() == COUNTERCLOCKWISE);
     const Polygon_with_holes_2 pwh1(pgn1);
     common_operator(pwh1, pgn2, outer_boundary, holes);
   }
 
 private:
+  //!
```
Comment (albert-github) [inline Minkowski_sum_2/include/CGAL/Minkowski_sum_2/Minkowski_sum_by_reduced_convolution_2.h:123]: Why this extra comment line?

**Comment by efifogel** [comment]
Cleaned up
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.6-Ic-156

**Comment by lrineau** [comment]
I will merge this pull-request in `master`. But the issue https://github.com/CGAL/cgal/issues/5830 was reported against CGAL-5.2.2, and is probably also present in the release branches we maintain (for CGAL-5.4.x and CGAL-5.5.x). Would it make sense to backport the patch?

**Pull request merged**